### ISSUE TITLE
Hyperparameters for replicating the results.

### ISSUE THREAD

**Issue by arjunsinghrathore** [vpariza/NeCo#2]
Hi,

I was trying to replicate the results from Table 1 of the paper and was trying out with the experiments/configs/neco_224x224.yml file provided in the repo. However, I am not fully able to replicate the results. So wanted to ask if there we need to set a different batch size or any other hyperparameter for replicating the results. Currently, I am using the 6 batch size, as provided in the config file with 1 GPU.

I am getting Semantic segmentation mIoU of 47 whereas it was 55.1 for NeCo as seen in Table 1b. 
And I am getting a Pascal VOC clustering with k=500 62 whereas it was 72.6 as seen in Table1a.

Confirming that we get the above two results in validation itself of the training loop of NeCo as provided in the code.

**Comment by vpariza** [comment]
Hi, many thanks to your question. I am happy to see people trying our code.

Now, in terms of your question in table 1b, we report unsupervised semantic segmentation which is described more thoroughly in appendix B.1. To evaluate it on the unsupervised semantic segmentation you need to use the code from [fully_unsup_seg](https://github.com/vpariza/NeCo/tree/main/experiments/fully_unsup_seg).

And yes indeed we used batch size of 64, so I assume that is the reason you got different results.

Hope I helped.

**Comment by arjunsinghrathore** [comment]
Thank you so much for the prompt response. I have a few quick follow-up questions

1. The attn_jacs_voc we get in the validation step of training does that correspond to what mIoU [fully_unsup_seg](https://github.com/vpariza/NeCo/tree/main/experiments/fully_unsup_seg)? 
2. The miou_kmeans we get in the validation step of training does corresponds to what we might get when we run the experiments/overcluster/eval_overcluster.py with k=500 cluters?
3. Lastly the when you say a batch size of 64, do you mean you set the batch_size in the config file as 64 or the effective batch size (bs * gpus * res_w ** 2) is 64? What is res_w here? The number of patches? 

Because the strange behaviour I am seeing is that on increasing the batch size I am getting worse attn_jacs_voc and miou_kmeans? So just want to confirm that if I should focus on the attn_jacs_voc and miou_kmeans I get in the validation step while training NeCo...or I need to separately run fully_unsup_seg.py and eval_overcluster.py to get the results reported in Table 1.

It would be really helpful if you could answer these questions. Thanks!

**Comment by vpariza** [comment]
1. No it is not the same as far as I remember. You need to run the specific [fully_unsup_seg](https://github.com/vpariza/NeCo/tree/main/experiments/fully_unsup_seg).
2. It corresponds to it but the `miou_kmeans` version of validation can be  configured to run for only a subset of validation iterations, and thus not going over the whole validation dataset. This is configured from the argument `val_iters` (https://github.com/vpariza/NeCo/blob/0e8cee1fdbdb11e3cbbd3776445aec8880a759e0/experiments/configs/neco_224x224.yml#L51) in the config files. More specifically if you had a validation dataset of 3200 images, and a validation batch size of 32, then the validation would need 100 itrations for a full validation. Now if you set `vals_iters` to 20 it will stop after the 20th iteration. You can set the `val_iters` variable to `null` and then it will be evaluate on the whole validation dataset though it will be also significantly slower. The check for the variable in the code is done here (https://github.com/vpariza/NeCo/blob/0e8cee1fdbdb11e3cbbd3776445aec8880a759e0/src/neco.py#L518).
3. I mean batch size as configured in the file. We experimented primarily on one gpu with 64 batch size, but I suppose it works on multiple gpus as well without any issue. `res_w` is the number of patches in one dimension (since the images are squares) .
I would say during training focus primarily on the `miou_kmeans`  (and potentially change the `val_iters`), and after training test with the [fully_unsup_seg](https://github.com/vpariza/NeCo/tree/main/experiments/fully_unsup_seg).

**Comment by arjunsinghrathore** [comment]
Thank you so much for the detailed response. 

I just noticed that in the [configure_optimizers function](https://github.com/vpariza/NeCo/blob/0e8cee1fdbdb11e3cbbd3776445aec8880a759e0/src/neco.py#L195) in neco.py, only the backbone_params_named is being used for populating with the params but head_params_named remains empty.....which means that backbone params and head params were not separated and all of them combined are in backbone_params_named...and while setting the adamw optimizer all the params are set with a learning rate of 1e-6 (lr_backbone)?

But in the NeCo paper it was mentioned that backbone and head params were trained with a separate learning rate?

**Comment by vpariza** [comment]
Hello again,

You are absolutely right. 

So, when I created this repository I adapted, simplified and removed parts of the paper's code. By mistake I forgot to include that part for the head parameters in the final version. 

I added it correctly into the new commit: https://github.com/vpariza/NeCo/commit/0ded54628bfc6d1abcc7b220b246d267f1dc9cec

I will go over again the code to make sure that everything else is there.

Now, whether it will affect the results too much, it will affect them but not too much. 

Thanks for pointing it out.
